### ISSUE TITLE
docker daemon panic " fatal error: concurrent map read and map write"

### ISSUE THREAD

**Issue by jinuxstyle** [docker/docker#29472]
**Description**

We just setup a swarm cluster with four nodes (3 managers and 1 worker) and haven't run our applications yet. There are only a few services and containers running. Only a couple of hours after the setup, I found the docker daemons on two nodes (both are manager nodes) are down. And then I checked the log, found the panic information. After that, I could still start the docker daemons without problem and they joined the swarm cluster. But I am afraid the panic would still happen sooner or later. 

**Steps to reproduce the issue:**
1. install docker version 1.13.0-rc2 on all nodes
2. setup swarm, 3 managers and 1 worker
3. several hours later, two docker daemons down

**Describe the results you received:**
Some docker daemons down, following is the log for one of the crashed docker.

Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: time="2016-12-15T22:12:15+08:00" level=info msg="Firewalld running: false"
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: fatal error: concurrent map read and map write
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: goroutine 665 [running]:
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: runtime.throw(0x198eda1, 0x21)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /usr/local/go/src/runtime/panic.go:566 +0x95 fp=0xc42f667988 sp=0xc42f667968
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: runtime.mapaccess2(0x172cae0, 0xc42020e7e0, 0xc42f667b20, 0x0, 0x0)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /usr/local/go/src/runtime/hashmap.go:340 +0x249 fp=0xc42f6679d0 sp=0xc42f667988
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).addServiceBinding(0xc4202f63c0, 0xc42abae654, 0x9, 0xc42573af40, 0x19, 0xc421327400, 0x19, 0xc42e02f1c0, 0x40, 0xc42abae6b0, ...)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/service_common.go:80 +0x174 fp=0xc42f667b50 sp=0xc42f6679d0
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).handleEpTableEvent(0xc4202f63c0, 0x1844740, 0xc42da4aff0)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:652 +0x523 fp=0xc42f667f20 sp=0xc42f667b50
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).(github.com/docker/docker/vendor/github.com/docker/libnetwork.handleEpTableEvent)-fm(0x1844740, 0xc42da4aff0)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:276 +0x3e fp=0xc42f667f48 sp=0xc42f667f20
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).handleTableEvents(0xc4202f63c0, 0xc422800d20, 0xc422d4a6f0)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:560 +0x82 fp=0xc42f667f88 sp=0xc42f667f48
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: runtime.goexit()
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /usr/local/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc42f667f90 sp=0xc42f667f88
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: created by github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).agentInit
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:276 +0x56f
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: goroutine 1 [chan receive, 172 minutes]:
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: main.(*DaemonCli).start(0xc42047b4d0, 0x0, 0x197b65f, 0x17, 0xc420465c00, 0xc420484320, 0xc4203c0ab0, 0x0, 0x0)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/cmd/dockerd/daemon.go:327 +0x183e
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: main.runDaemon(0x0, 0x197b65f, 0x17, 0xc420465c00, 0xc420484320, 0xc4203c0ab0, 0x0, 0x0)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/cmd/dockerd/docker.go:86 +0xb2
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: main.newDaemonCommand.func1(0xc42048e000, 0x2525ab8, 0x0, 0x0, 0x0, 0x0)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/cmd/dockerd/docker.go:42 +0x71
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc42048e000, 0xc42000c260, 0x0, 0x0, 0xc42048e000, 0xc42000c260)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:646 +0x26d
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42048e000, 0x176d340, 0xc420484301, 0xc42032b290)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:742 +0x377
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42048e000, 0xc42032b290, 0xc420082058)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:695 +0x2b
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: main.main()
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /go/src/github.com/docker/docker/cmd/dockerd/docker.go:106 +0xe2
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: goroutine 17 [syscall, 172 minutes, locked to thread]:
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: runtime.goexit()
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /usr/local/go/src/runtime/asm_amd64.s:2086 +0x1
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: goroutine 6 [syscall, 164 minutes]:
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: os/signal.signal_recv(0x24cb580)
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: /usr/local/go/src/runtime/sigqueue.go:116 +0x157
Dec 15 22:12:15 node2.manager.oa1.dkcloud dockerd[25200]: os/signal.loop()

**Describe the results you expected:**
No docker daemons down

**Additional information you deem important (e.g. issue happens only occasionally):**
I have setup two swarm clusters, so far I only saw the issue on one of the cluster which has more network setups (like ovs) than the other.

**Output of `docker version`:**

```
 Version:      1.13.0-rc2
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   1f9b3ef
 Built:        Wed Nov 23 17:40:58 2016
 OS/Arch:      linux/amd64

Server:
 Version:             1.13.0-rc2
 API version:         1.25
 Minimum API version: 1.12
 Go version:          go1.7.3
 Git commit:          1f9b3ef
 Built:               Wed Nov 23 17:40:58 2016
 OS/Arch:             linux/amd64
 Experimental:        false

```

**Output of `docker info`:**

```
# docker info
Containers: 51
 Running: 1
 Paused: 0
 Stopped: 50
Images: 5
Server Version: 1.13.0-rc2
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: t2lhotbf8u8ke9pcxzi224pvq
 Is Manager: true
 ClusterID: lb9blnw029fzyl1edlfl545q7
 Managers: 3
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.0.40
 Manager Addresses:
  0.0.0.0:2377
  192.168.0.40:2377
  192.168.0.42:2377
  192.168.0.43:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 51371867a01c467f08af739783b8beafc154c4d7
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.2-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 24
Total Memory: 31.41 GiB
Name: node2.manager.oa1.dkcloud
ID: ANRI:F73L:IUI4:A45M:HQKR:IFLK:KW4G:64SF:MMWD:TRYN:776I:WWEL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 1995
 Goroutines: 3582
 System Time: 2016-12-16T11:01:37.613911055+08:00
 EventsListeners: 3
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Labels:
 label_host=192.168.0.40:2375
 label_node=192.168.0.40
Experimental: false
Insecure Registries:
 192.168.0.44:5000
 192.168.0.44:5001
 127.0.0.0/8
Live Restore Enabled: false

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by mavenugo** [comment]
ping @sanimej @aboch is this addressed with one of the concurrency fixes post in RC3 & RC4 ?

**Comment by jinuxstyle** [comment]
I searched the error message and found a similar issue #28237. From their logs, they both got to the same panic point but from different path, one from volume/store and the other from libnetwork. Hope this info is helpful.

**Comment by aboch** [comment]
@jinuxstyle The fix for this issue is in docker 1.13.0-rc3

**Comment by jinuxstyle** [comment]
@aboch That's great. :) I will upgrade to 1.13.0-rc3. Thanks.
